source/agora/consensus/data/Block.d

```diff
@@ -824,3 +824,53 @@ version (unittest)
         return block;
     }
 }
+
+/// demonstrate signing two blocks at height 1 to reveal private node key
+unittest
+{
+    import agora.consensus.data.genesis.Test: GenesisBlock;
+    import agora.crypto.ECC: Scalar, Point;
+    import agora.crypto.Schnorr: multiSigSign, multiSigVerify, Pair, Sig;
+    import std.format;
+
+    Hash random_seed1 = "seed1".hashFull();
+    Hash random_seed2 = "seed2".hashFull();
+
+    // Generate two blocks at height 1
+    auto block1 = GenesisBlock.makeNewBlock(
+        genesisSpendable().take(2).map!(txb => txb.sign()).array(), 10, random_seed1);
+    auto block2 = GenesisBlock.makeNewBlock(
+        genesisSpendable().take(3).map!(txb => txb.sign()).array(), 10, random_seed2);
+
+    Scalar v = genesis_validator_keys[0].secret;
+    const Scalar rc = Scalar(hashMulti(v, "consensus.signature.noise", 0));
+    const Scalar r = rc + Scalar(block1.header.height.hashFull());
+
+    // Two messages
+    Scalar c1 = block1.hashFull();
+    Scalar c2 = block2.hashFull();
+    assert(c1 != c2);
+
+    // Sign with same r twice
+    Scalar s1 = multiSigSign(r, v, c1);
+    Scalar s2 = multiSigSign(r, v, c2);
+
+    // Verify signatures
+    assert(multiSigVerify(Sig(r.toPoint(), s1), v.toPoint(), c1));
+    assert(multiSigVerify(Sig(r.toPoint(), s2), v.toPoint(), c2));
+
+
+    // Calculate the private key
+    // s = r + (v * c)
+    // s1 - s2 = r + (v * c1) - (r + v * c2) = v(c1 - c2)
+    // v = (s1 - s2) / (c1 - c2)
+    Scalar s = (s1 - s2);
+    assert(s1 == s + s2);
+    Scalar c = (c1 - c2);
+    assert(c1 == c + c2);
+
+    Scalar secret = s * c.invert();
+    assert(secret == v,
```

Comment:
I can not figure out why this is not working.

Reply:
I'll take a deeper look at this tomorrow.

Reply:
Where did you get this formula from?

Reply:
https://tlu.tarilabs.com/cryptography/digital_signatures/introduction_schnorr_signatures.html#replay-attacks

Reply:
Also https://en.wikipedia.org/wiki/Schnorr_signature